### PR TITLE
JDK-8304723: Statically allocate global mutexes

### PR DESCRIPTION
**Pull request by jcking**
Switch to statically allocate storage for all of Hotspot's global mutexes and monitors. This avoids needing to allocate each one dynamically at startup. These also live for the lifetime of the program, so this reduces any potential memory fragmention in the C heap.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issues
 * [JDK-8304723](https://bugs.openjdk.org/browse/JDK-8304723): Statically allocate global mutexes
 * [JDK-8304736](https://bugs.openjdk.org/browse/JDK-8304736): Heap_lock is created twice

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [7491f771](https://git.openjdk.org/jdk/pull/13143/files/7491f771fac0dca094bac0e5e53afdcf6a4da798)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to [7491f771](https://git.openjdk.org/jdk/pull/13143/files/7491f771fac0dca094bac0e5e53afdcf6a4da798)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13143/head:pull/13143` \
`$ git checkout pull/13143`

Update a local copy of the PR: \
`$ git checkout pull/13143` \
`$ git pull https://git.openjdk.org/jdk.git pull/13143/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13143`

View PR using the GUI difftool: \
`$ git pr show -t 13143`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13143.diff">https://git.openjdk.org/jdk/pull/13143.diff</a>

</details>

### CODE REVIEW

**Comment by jcking** [comment]
Or something like `ReentrantMutexLocker`, since its technically making sure not to re-lock a mutex the calling thread already owners.

**Comment by jcking** [comment]
This has also surfaced two existing bugs:  `UnregisteredClassesTable_lock` is never created and always `nullptr` and `Heap_lock` is created twice.

@dholmes-ora I assume I should fix those two issues in a separate PR. Should `UnregisteredClassesTable_lock` be `nosafepoint`?

**Comment by tstuefe** [comment]
Makes sense, but I think you can do a large part of what you want with a fraction of the changes.

- we can probably switch from dyn. allocating the mutex name to keeping it inline as a static buffer. Cuts allocations down by half already. The name lengths are not that surprising, and even if we truncate (we shouldn't) it's probably fine.
- On posix, switch from pthread_cond_init to static initialization if possible.

Other than that, moving initialization to static now runs the initializers at dll load time, not at VM initialization time. Should be fine, but it's a semantic change to be aware of.

Like Coleen I don't like the invasiveness of this patch, esp. considering what it is doing. Pls try to simplify. You could probably keep the outside semantics, and do without templates. Eg define the mutexes in a compilation unit and export pointers to them to the outside?

e.g. 
x.cpp
static whatever-mutex-type Currentname_obj;
extern whatever-mutex-type* Currentname = &Currentname_obj;

and the hpp file accordingly.

**Comment by jcking** [comment]
> Like Coleen I don't like the invasiveness of this patch, esp. considering what it is doing. Pls try to simplify. You could probably keep the outside semantics, and do without templates. Eg define the mutexes in a compilation unit and export pointers to them to the outside?

I did that initially, but it still has the unnecessary indirection potentially causing more cache misses. Additionally as I mentioned above, this has now caught two existing bugs where just raw points would not have caught them, since a nullptr is perfectly valid for `MutexLocker`. It doesn't catch the fact that the lock was never created. `StaticLock` asserts that the lock has been initialized when its accessed and can never return `nullptr`.  It also ensures it is never initialized twice, which `Heap_lock` was.

**Comment by jcking** [comment]
> * we can probably switch from dyn. allocating the mutex name to keeping it inline as a static buffer. Cuts allocations down by half already. The name lengths are not that surprising, and even if we truncate (we shouldn't) it's probably fine.

This is even more invasive and should be done separately, I already noticed it. There are places where mutex/monitor are not statically allocated and we have to ensure those are also compatible with that change.

> * On posix, switch from pthread_cond_init to static initialization if possible.

The pthread_cond_init static initialization is not useful here, we are not initializing during library loading. Everything is still done during VM initialization. Trying to use pthread_cond_init static initialization would be more complex as we would then have to introduce a tag type to select which one, as one is not normally supposed to destroy statically initialized condition variables.

**Comment by dholmes-ora** [comment]
> On posix, switch from pthread_cond_init to static initialization if possible.

Some platforms have the "wrong" default which is why we don't attempt to use the pthread static initializers. But in general statically initializing our Mutex/Monitor would be very fragile IMO - as discussed in the past.

Doing the allocation statically is a different matter. But what does this gain? Is the startup improvement noticeable?

> Created [JDK-8304736](https://bugs.openjdk.org/browse/JDK-8304736) and [JDK-8304738](https://bugs.openjdk.org/browse/JDK-8304738).

Thanks for that.

**Comment by tstuefe** [comment]
> > On posix, switch from pthread_cond_init to static initialization if possible.
> 
> Some platforms have the "wrong" default which is why we don't attempt to use the pthread static initializers. But in general statically initializing our Mutex/Monitor would be very fragile IMO - as discussed in the past.

Oh, right, there was that. Wonder if that still holds true today. Wasn't that Solaris?

> 
> Doing the allocation statically is a different matter. But what does this gain? Is the startup improvement noticeable?
> 

I'd be very happy if startup would be fast enough for ~200 fewer mallocs to matter. 

A blank hello world with no args to speak of calls into malloc >10k times. ~150(?) mallocs less are a step in the right direction.

My major gripe with this change is that I don't like the SimpleLock solution and suspect this could be simplified. I don't like the changes to the call sites either. 

A simple change to keep the mutex name in-line instead of mallocing it separately would already half the number of mallocs, and that would be a really trivial patch. One step further, one could allocate the mutexes in an array, or inside a single holder object. There are probably more ways to do this.

**Comment by tstuefe** [comment]
In case I was not clear: I like the intention of this patch and would prefer static allocation for Mutexes too.

**Comment by jcking** [comment]
> I'd be very happy if startup would be fast enough for ~150 fewer mallocs to matter.
> 
> A blank hello world with no args to speak of calls into malloc >10k times. ~150(?) mallocs less are a step in the right direction.

Baby steps. :)

As a compromise I'll shift to keeping the pointers and just use a static storage variable in the `def` macros, as suggested. It accomplishes the goal if not malloc'ing but is not as safe due to the two aforementioned bugs. I'll also add an assert to catch the "created twice" problem. But having the mutex pointer always be nullptr is a different story that cannot be caught with that approach.

**Comment by tstuefe** [comment]
> > I'd be very happy if startup would be fast enough for ~150 fewer mallocs to matter.
> > A blank hello world with no args to speak of calls into malloc >10k times. ~150(?) mallocs less are a step in the right direction.
> 
> Baby steps. :)
> 
> As a compromise I'll shift to keeping the pointers and just use a static storage variable in the `def` macros, as suggested. It accomplishes the goal if not malloc'ing but is not as safe due to the two aforementioned bugs. I'll also add an assert to catch the "created twice" problem. But having the mutex pointer always be nullptr is a different story that cannot be caught with that approach.

The create twice problem is admittedly evil. The nullptr problem not so much, since we will access it anyway and crash right there, or?

**Comment by jcking** [comment]
> > > I'd be very happy if startup would be fast enough for ~150 fewer mallocs to matter.
> > > A blank hello world with no args to speak of calls into malloc >10k times. ~150(?) mallocs less are a step in the right direction.
> > 
> > 
> > Baby steps. :)
> > As a compromise I'll shift to keeping the pointers and just use a static storage variable in the `def` macros, as suggested. It accomplishes the goal if not malloc'ing but is not as safe due to the two aforementioned bugs. I'll also add an assert to catch the "created twice" problem. But having the mutex pointer always be nullptr is a different story that cannot be caught with that approach.
> 
> The create twice problem is admittedly evil. The nullptr problem not so much, since we will access it anyway and crash right there, or?

Unfortunately not. It's used with MutexLocker which happily accepts `nullptr` and just doesn't do anything. That is a feature of it. So if you fail to create a mutex, you might not notice and the data being guarded by the mutex is no longer being accessed safely.

**Review by coleenp — APPROVED** [review]
Ok, this is tricky but better than the more invasive change before.

File: src/hotspot/share/runtime/mutexLocker.cpp [hunk 1]
```diff
@@ -201,183 +195,187 @@ static void add_mutex(Mutex* var) {
...
 }
+#define MUTEX_DEFN(name, type, pri, ...) MUTEX_DEF(name, type, Mutex::pri, ##__VA_ARGS__)
 
 // Specify relative ranked lock
 #ifdef ASSERT
-#define defl(var, type, held_lock, ...) {         \
-  var = new type(held_lock->rank()-1, #var, ##__VA_ARGS__); \
-  add_mutex(var);                                      \
-}
+#define MUTEX_DEFL(name, type, held_lock, ...) MUTEX_DEF(name, type, (held_lock)->rank()-1, ##__VA_ARGS__)
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/mutexLocker.cpp:210]: Pre-existing but what does the 'L' suffix mean here? We now have _DEF _DEFN and _DEFL and I'm not clear what the N and L signify. ??

Nit while you are here `rank()-1` -> `rank() - 1`

**Review by dholmes-ora — APPROVED** [review]
This version seems fine.

One query below.

Thanks.

**Comment by tstuefe** [comment]
@kimbarrett

> 
> I see no mallocs for the names. The names come from macro stringizing the associated variable in defl(), so are string literals.
> 
> The names used to be embedded in Mutex (providing some padding around the atomic parts I think), with the string passed to the constructor strcpy?ed into that embedded buffer, but that was removed by JDK-8264146.
> 

Names are strduped since https://bugs.openjdk.org/browse/JDK-8265682.
https://github.com/openjdk/jdk/blob/7491f771fac0dca094bac0e5e53afdcf6a4da798/src/hotspot/share/runtime/mutex.cpp#L277

**Review by tschatzl — APPROVED** [review]
Lgtm.

Please close https://bugs.openjdk.org/browse/JDK-8304736 again now that this issue has been fixed, or probably better use `/issue add JDK-8304736`

File: src/hotspot/share/runtime/mutexLocker.cpp (hunk 1 above)
Comment (jcking) [reply src/hotspot/share/runtime/mutexLocker.cpp]: Fixed the spacing to be `rank() - 1`.

As far as `N` and `L` I don't think its anything special. I chose `_DEF` to be the shared macro implementation and `N` as a short for `DEFiNition`. `L` I assume means lock, as in define using another lock's rank. But that is a guess.

**Comment by tstuefe** [comment]
@jcking looks a lot better, thank you!

You already have enough reviews. Gave it a cursory glance and it looks good to me.

Cheers, Thomas

File: src/hotspot/share/runtime/mutexLocker.cpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/share/runtime/mutexLocker.cpp]: Yes, the 'l" in defl was for "lock" originally.

**Review by dholmes-ora — APPROVED** [review]

**Comment by jcking** [comment]
GHA failure seems to be network issue on the underlying install dependencies step, so unrelated.

**Pull request closed without merging**